client/src/main/java/io/pravega/client/netty/impl/RawClient.java

```diff
@@ -131,12 +140,31 @@ private void closeConnection(Throwable exceptionToInflightRequests) {
     }
 
     public <T extends Request & WireCommand> CompletableFuture<Reply> sendRequest(long requestId, T request) {
+        if (this.isTesting) {
+            return sendRequest(requestId, request, Duration.ofSeconds(3600));
+        } else {
+            return sendRequest(requestId, request, Duration.ofSeconds(30));
+        }
+    }
+
+    public <T extends Request & WireCommand> CompletableFuture<Reply> sendRequest(long requestId, T request,
+                                                                                  Duration timeout) {
         return connection.thenCompose(c -> {
             log.debug("Sending request: {}", request);
             CompletableFuture<Reply> reply = new CompletableFuture<>();
             synchronized (lock) {
                 requests.put(requestId, reply);
             }
+
+            Futures.onTimeout(reply, timeout, connectionFactory.getInternalExecutor(), ex -> {
```

Comment:
This is very costly, we are adding lot of work and a few memory allocations on the hottest path.

What about saving just a deadline timestamp for the request in a map and periodically check the deadline and fail the request?

Reply:
Thanks. Yes this is kind of costly and I have discussed with Tom about this issue, it seems that rawclient is not on THE hottest path, but I think your way is more efficient.

Reply:
RawClient may not be on the "hottest" path, but it is used a lot for non-streaming purposes. It is used for KV Tables, for the Controller Metadata requests, etc. So we do owe it to ourselves to keep it as efficient as possible.

Reply:
You can 1) do that (store all requests in a map - there already is one) and periodically loop through it and verify or 2) consider using `Futures.futureWithTimeout` (set `CompletableFuture<Reply> reply = Futures.futureWithTimeout(duration, executor)`).

You should weigh the pros and cons of each of them. The thing I dislike about option 1 is that you will periodically have to acquire the lock to go through that map and inspect all outstanding requests. This could be introducing "cleanup pauses" in the client, which may be undesirable.

Reply:
@eolivelli After discussing with Andrei, we are a little uncertain about the performance cost you mentioned, Andrei has used this kind of technique extensively in segment store and this method doesn't show on the flamegraph. Perhaps we missed something and it would great if you could clarify the performance cost. 
Thanks

Reply:
If this is not one of the hottest paths than maybe there is no need to make it more complex.
I was just talking about saving memory allocations but if your analysis doesn't show that this is an hotspot then let's keep it simple